### PR TITLE
Handle timezones in bulk uploader and Fhir converter

### LINKED ISSUE #6025: Handle time zones in CSV uploader
### Description

When local datetimes in the format of `M/D/YYYY HH:mm` are converted from CSV, the datetime in FHIR should accurately reflect that time zone.

#### Example
`6/22/2023 06:00` is unclear whether it refers to the moment of time at 2pm Eastern or 2pm Pacific.

Currently, the datetime uses the system's default time zone (typically UTC) such that it picks the moment of time where UTC was 06:00. So a CSV input from local Pacific Time of `6/22/2023 06:00` ends up being converted to `6/21/2023 23:00` since Pacific Daylight Time is 7 hours behind UTC.

### Acceptance Criteria
- Local date times in a CSV file are converted to a zoned datetime value
- Users may add a timezone code as a suffix on date fields.
- If they do not add a time, we'll default to 12pm on the date entered
- If they do not enter a timezone, we'll default to the timezone of the testing lab address.
- If the testing lab address cannot be verified, we'll use `US/Eastern` as the default fallback
- This validation should happen when the other fields are also validated (before it starts the actual FHIR conversion)

Acceptable Datetime Field Formats
- `M/D/YYYY`
- `M/D/YYYY HH:mm`
- `M/D/YYYY HH:mm TZ`

Acceptable Values for TZ
- Any US specific timezone abbreviation such as ET, EST, EDT, CT, MST, PDT, AKST, HST.
- Any `TZ Identifier` defined by [official ICANN tz database](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones).

### PR DESCRIPTION
**Pull request by mpbrown**
# BACKEND PULL REQUEST

## Related Issue

- Resolves #6025 and #5987
- Related to #5844, #5946, and #5982 

## Changes Proposed

- Convert datetime strings from CSV to ZonedDateTimes so that datetimes in FHIR bundles refer to the correct instant of time
    - For datetimes that are not associated with any timezone, use UTC time with `Z` suffix.
        - Example: a timestamp of when an entry was updated
        - Format: `2021-12-20T19:00:00Z`
    - For datetimes that are associated with a timezone, use the UTC offset.
        - Example: datetime of specimen collection
        - Format: `2021-12-20T14:00:00-05:00`
- Updates DateTime regex to allow optional timezone code
    - Some US codes like ET, EST, EDT, CT, MST, PDT, AKST, HST, etc are allowed
    - Any `TZ identifier` string defined by [ICANN tz database](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones)
- If timezone is not specified in the datetime, use the relevant address to retrieve [timezone metadata](https://www.smarty.com/docs/cloud/us-street-api#timezone) from Smarty.
- If address validation fails, use fallback timezone of `US/Eastern`
- Updates unit tests and removes now obsolete code in `FhirConverter` and `BulkUploadResultsToFhir`

## Additional Information

Before we can convert it to a UTC timestamp, we need to convert the local `test_result_date` to the correct Date instant of time which requires knowing the data's time offset from UTC and whether it obeys daylight savings.

This PR attempts to validate the address of the testing lab and use the [time zone metadata](https://www.smarty.com/docs/cloud/us-street-api#timezone) returned from Smarty for the correct UTC offset.

## Testing

- Check updated unit tests
- Test on dev7 to make sure both the bulk uploader and single entry FHIR conversions work as expected

### CODE REVIEW

**Comment by mpbrown** [comment]
Created a new issue to track the time zone work specifically of referring to the bulk uploader test coverage issue. #6025 

Will push some additional changes based on this soon

**Review by BobanL — APPROVED** [review]
Changes look good and worked as expected in dev7! Thanks you 😄

**Review by BobanL — APPROVED** [review]

**Review by nathancrtr — DISMISSED** [review]
This looks great! 👏 thanks for the well-written PR notes and unit tests.

**Review by zdeveloper — CHANGES_REQUESTED** [review]
partial review

File: backend/src/main/java/gov/cdc/usds/simplereport/api/converter/FhirConverter.java [hunk 1]
```diff
@@ -788,32 +812,47 @@ public DiagnosticReport convertToDiagnosticReport(TestEvent testEvent) {
               testEvent.getDeviceType().getSupportedDiseaseTestPerformed());
     }
 
+    ZonedDateTime dateTested = null;
+    if (testEvent.getDateTested() != null) {
+      var zoneId =
+          addressValidationService.getZoneIdByAddress(testEvent.getFacility().getAddress());
```
[the remark this answers is not in the document]
Comment (mpbrown) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/converter/FhirConverter.java:818]: Good catch for caching!

For that particular method overload of `convertToDiagnosticReport(TestEvent testEvent)`, is that being used for single entry conversion to FHIR? I tried following the usage up to where the reporting service beans are defined but still wasn't sure how we should handle the TestEvent's `dateTested`. Can we expect that field from the database to be a UTC datetime?

[the remark this answers is not in the document]
Comment (zdeveloper) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/converter/FhirConverter.java:818]: from what I can see, `convertToDiagnosticReport` is only used in the bulk uploader to FHIR.

Also everything that's saved to the db is converted to UTC from the clients machine in the single entry flow

**Comment by mpbrown** [comment]
Updated with the new acceptance criteria in #6025 as well as caching and additional test coverage

Dev7 is freshly deployed and smoked tested

**Comment by nathancrtr** [comment]
@mpbrown are the datetime format requirements from [the ticket](https://github.com/CDCgov/prime-simplereport/issues/6025) up-to-date?

Screen Shot 2023-07-06 at 4 35 41 PM

If so, I'm encountering a problem with (at least) the `M/D/YYYY HH:mm TZ` format. When using `12/20/2021 14:00 PDT` as the value, the upload fails with the following errors:

Screen Shot 2023-07-06 at 4 39 56 PM

**Comment by mpbrown** [comment]
> @mpbrown are the datetime format requirements from [the ticket](https://github.com/CDCgov/prime-simplereport/issues/6025) up-to-date?
> 
> Screen Shot 2023-07-06 at 4 35 41 PM
> If so, I'm encountering a problem with (at least) the `M/D/YYYY HH:mm TZ` format. When using `12/20/2021 14:00 PDT` as the value, the upload fails with the following errors:
> 
> Screen Shot 2023-07-06 at 4 39 56 PM

Nathan and I messaged about this in Slack and we may have to check with ReportStream to see what datetime formats they consider valid.

Test data csv includes:
- Specimen_collection_date - `12/21/2021 14:00 PDT`
- Testing_lab_specimen_received_date - `12/22/2021 14:00 PDT`

FHIR bundle gets created successfully with the following Specimen entry, correctly parsing the datetime strings as Pacific Time with UTC offset during December as -8

```
{
    "fullUrl": "Specimen/326117e3-f4f5-44cc-9387-407cd3601e44",
    "resource": {
        "resourceType": "Specimen",
        "id": "326117e3-f4f5-44cc-9387-407cd3601e44",
        "identifier": [
            {
                "value": "367f31bf-3bf2-43cd-8be4-dcab1daeb256"
            }
        ],
        "subject": {
            "reference": "Patient/1234"
        },
        "receivedTime": "2021-12-22T14:00:00-08:00",
        "collection": {
            "collectedDateTime": "2021-12-21T14:00:00-08:00"
        }
    }
},
```

**Comment by mpbrown** [comment]
Added a data transformation step for the covid csv pipeline that converts the user's uploaded datetime strings to an offset format that is accepted by ReportStream

Re-deployed on [dev7](https://dev7.simplereport.gov/app/) with a successful ReportStream response

File: backend/src/main/java/gov/cdc/usds/simplereport/service/errors/InvalidAddressException.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package gov.cdc.usds.simplereport.service.errors;
+
+import java.io.Serial;
+
+public class InvalidAddressException extends RuntimeException {
+  @Serial private static final long serialVersionUID = 1L;
```
Comment (zdeveloper) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/errors/InvalidAddressException.java:6]: I am not familiar with @Serial, what does it do?

File: backend/src/main/java/gov/cdc/usds/simplereport/utils/BulkUploadResultsToFhir.java [hunk 3]
```diff
@@ -56,12 +55,11 @@
 public class BulkUploadResultsToFhir {
 
   private static final String ALPHABET_REGEX = "^[a-zA-Z\\s]+$";
-  private static final String SNOMED_REGEX = "(^[0-9]{9}$)|(^[0-9]{15}$)";
-  private final ResultsUploaderDeviceValidationService resultsUploaderDeviceValidationService;
+  private static final String SNOMED_REGEX = "(^\\d{9}$)|(^\\d{15}$)";
```
Comment (zdeveloper) [inline backend/src/main/java/gov/cdc/usds/simplereport/utils/BulkUploadResultsToFhir.java:58]: whats the reason for this change?

File: backend/src/test/java/gov/cdc/usds/simplereport/test_util/JsonTestUtils.java [hunk 4]
```diff
@@ -75,7 +77,8 @@ public static void assertJsonNodesEqual(
               fieldPath + "[" + i + "]");
         }
       } else { // Otherwise assert values are equal
-        assertThat(actualFieldValue.toString()).isEqualTo(expectedFieldValue.toString());
+        log.debug("Comparing " + fieldPath);
```
Comment (zdeveloper) [inline backend/src/test/java/gov/cdc/usds/simplereport/test_util/JsonTestUtils.java:80]: do we need this?

**Review by zdeveloper — APPROVED** [review]
LGTM! thank you for all this work

File: backend/src/main/java/gov/cdc/usds/simplereport/service/errors/InvalidAddressException.java (hunk 2 above)
Comment (mpbrown) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/errors/InvalidAddressException.java:6]: That particular `@Serial` annotation was recommended by Intellij to allow compile-time checking of serialization-related declarations, similar to the checking enabled by the `@Override` annotation type to validate method overriding

File: backend/src/main/java/gov/cdc/usds/simplereport/utils/BulkUploadResultsToFhir.java (hunk 3 above)
Comment (mpbrown) [reply backend/src/main/java/gov/cdc/usds/simplereport/utils/BulkUploadResultsToFhir.java:58]: SonarLint complaining that the digits could be replaced by `\d`, I could revert this if we prefer it to explicitly still use [0-9]

File: backend/src/test/java/gov/cdc/usds/simplereport/test_util/JsonTestUtils.java (hunk 4 above)
Comment (mpbrown) [reply backend/src/test/java/gov/cdc/usds/simplereport/test_util/JsonTestUtils.java:80]: Useful when debugging the expected and actual JSON output for the FHIR bundler, but not necessary to keep in so I could comment out or remove

**Review by nathancrtr — APPROVED** [review]
I re-tried my test file in dev7 and the upload succeeded as expected. Thanks for investigating that issue! This LGTM 🚀

**Pull request merged**